src/Featurizers/Components/StatisticsMetricsEstimator.h

```diff
@@ -0,0 +1,515 @@
+    // ----------------------------------------------------------------------
+    StandardStatisticalAnnotaionData(typename TypeSelector::SumTypeSelector<T>::type sum, std::double_t average, T min, T max, std::uint64_t count);
+    ~StandardStatisticalAnnotaionData(void) = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(StandardStatisticalAnnotaionData);
+};
+
+namespace Details {
+/////////////////////////////////////////////////////////////////////////
+///  \class         BasicStatsUpdater
+///  \brief         BasicStatsUpdater updates basic stats include
+///                 min, max and count
+///
+template <typename T>
+class BasicStatsUpdater {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using InputType                         = T;
+
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    // first_element_flag is default to be true in constructor to guarantee
+    // min, max and count are set based on first valid input element
+    BasicStatsUpdater(void);
+
+    void update(InputType input);
+    // commit returns the result of min, max and count
+    std::tuple<InputType, InputType, std::uint64_t> commit(void);
+protected:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Protected Data
+    // |
+    // ----------------------------------------------------------------------
+    InputType                                 _min;
+    InputType                                 _max;
+    std::uint64_t                             _count;
+    bool                                      _first_element_flag;
+};
+/////////////////////////////////////////////////////////////////////////
+///  \class         StandardStatsTrainingOnlyPolicy
+///  \brief         StandardStatsTrainingOnlyPolicy updates advanced stats include
+///                 sum and average
+///
+template <typename T>
+class StandardStatsUpdater : public BasicStatsUpdater<T> {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using InputType                         = T;
+
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    StandardStatsUpdater(void);
+
+    void update(InputType input);
+    // commit returns the result of min, max, count, sum and average
+    std::tuple<typename TypeSelector::SumTypeSelector<T>::type, std::double_t, InputType, InputType, std::uint64_t> commit(void);

```

Comment:
Return a struct

Reply:
I feel like I can directly return AnnotationData for basic and standard since a struct would be exactly the same as AnnotationData.

Reply:
I would recommend implementing AnnotationData in terms of these new structs.